{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ae154a92-a5a7-4ff7-a98c-f610e30b7939",
   "metadata": {},
   "source": [
    "# Data\n",
    "\n",
    "For hands-on activity, we will make use a part of the dataset generated by the following study:\n",
    "\n",
    "Metagenomic Analysis of the Abundance and Composition of Antibiotic Resistance Genes in Hospital Wastewater in Benin, Burkina Faso, and Finland. Markkanen MA, Haukka K, Pärnänen KMM, Dougnon VT, Bonkoungou IJO, Garba Z, Tinto H, Sarekoski A, Karkman A, Kantele A, Virta MPJ. mSphere 8(1): e0053822 (2023 Feb)\n",
    "\n",
    "The full dataset is available here: https://www.ebi.ac.uk/ena/browser/view/PRJEB47975 . \n",
    "\n",
    "For this workshop, we will use only 8 of the following (heavily-downsampled) samples.\n",
    "\n",
    "| Country |  Sample ID | ENA Accession ID |\n",
    "| ------- | ---------- | ---------------- | \n",
    "| Finland | FH1 | ERR7015395 | \n",
    "| Finland | FH2 | ERR7015396 |\n",
    "| Finland | FH3 | ERR7015397 | \n",
    "| Finland | FH4 | ERR7015398 |\n",
    "| Benin   | BH1 | ERR7015311 |\n",
    "| Benin   | BH2 | ERR7015312 |\n",
    "| Benin | BH3 | ERR7015313 |\n",
    "| Benin | BH4 | ERR7015315 |\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad88a8b8-cf74-4093-ac89-52085c7a3697",
   "metadata": {},
   "source": [
    "# Data Sneak Peek\n",
    "If you followed the instructions in the Section [Getting Started](getting-started.md) correctly, you should be able to see the files by doing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1afe8e35-8739-49d9-87eb-5aad588c2355",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "BH1_1.fastq.gz\tBH3_1.fastq.gz\tFH1_1.fastq.gz\tFH3_1.fastq.gz\tmetadata.txt\n",
      "BH1_2.fastq.gz\tBH3_2.fastq.gz\tFH1_2.fastq.gz\tFH3_2.fastq.gz\n",
      "BH2_1.fastq.gz\tBH4_1.fastq.gz\tFH2_1.fastq.gz\tFH4_1.fastq.gz\n",
      "BH2_2.fastq.gz\tBH4_2.fastq.gz\tFH2_2.fastq.gz\tFH4_2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "! ls data/metagenome_samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95f194d8-7969-4efb-a3c5-4ff0aab21768",
   "metadata": {},
   "source": [
    "The sequence data is in `fastq' format, and for each sample, there is a pair of fastq files, e.g. BH1_1.fastq and BH1_2.fastq. The pair comes from the fact that when a long DNA molecule is sequenced, it is first fragmented into pieces, and each piece is read from two ends. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe6102d0-12e9-4c2f-accb-7d64f7f0c8a6",
   "metadata": {},
   "source": [
    "You can have a look inside one of the files by opening up a terminal (click the blue + button on the left top and click on Terminal) and run the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "29d15bf9-8721-4fbc-a141-6e98160b58f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@ERR7015311.1 A00464:196:HFFTTDSXY:2:1101:10004:15687/1\n",
      "GGCCATGTCGGCGCGCTCGGCGGGGAACTCGCGGATGTCGGCGCCGGCGGCGAAGTTGCCTCCCTCGCCGCGCACGATCACGCAGCGCAGCGCATTGTCGGCCGCCAGCTGGTCGAACACCGCGCGCAGCTCGCCCCACATGCCCACGGTG\n",
      "+\n",
      "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF,FFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFF\n",
      "@ERR7015311.2 A00464:196:HFFTTDSXY:2:1101:10004:16063/1\n",
      "GCCAGCATCGCGAGCTCGGCTTCGACCACGGCGATGTCGGCCTTGTCCTTGGCGTTGCGCAGCGCCTCTTCAGCGCGCTGGCGCGCTTCCAGGGCACGGGCCTCGTCCAGGTCGGCGGCACGGATGGCCGTGTCGGCCAGGACCGTGACGC\n",
      "+\n",
      "FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF:FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF\n",
      "@ERR7015311.3 A00464:196:HFFTTDSXY:2:1101:10004:18505/1\n",
      "ACGCGGTGTTCGAAGGCGCGGTCATCGCCGTGGCGGCGCATCACCTTTATCCCGCCGTGCCGTACTGGCTGGTGGCGCTGGCGGTGGTGATTTACAGCGTGCTGCTGATCTTCGGCAGCGTGCAGCGCTGGTTGGACAAGTTCAACGGCGT\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "! zcat data/metagenome_samples/BH1_1.fastq.gz | head "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a978700-8081-4897-99dd-71192ab2606e",
   "metadata": {},
   "source": [
    "A block of four lines correspond to one sequence entry often called a `read'. The first line starting with @ contains the read ID. The next line contains the actual sequence. In the example above, it is 100 nucleotides long. The fourth line provides information about the quality of the estimate of the charaater at each position of the read. \n",
    "There are these many files in one file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d80c652d-a933-417e-bcff-d27530e85cc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4000000\n"
     ]
    }
   ],
   "source": [
    "! zcat data/metagenome_samples/BH1_1.fastq.gz | wc -l"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9035da8b-dbed-4df1-9b38-e25ee06c86d8",
   "metadata": {},
   "source": [
    "So that means, there are 1 million reads in one file.\n",
    "\n",
    "The original dataset has 54.9 million reads as can be seen here https://www.ebi.ac.uk/ena/browser/view/ERR7015311.  \n",
    "\n",
    "As you can see, we have drastically downsampled for this workshop."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a3e00bfc-b911-4b94-8cec-f241e7dc7411",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
