<!-- # The Data Science of Metagenomics : Bioinformatics Workflows for Profiling Microbial Communities -->
# The Bioinformatics of Profiling Microbial Communities in Wastewater from Shotgun Metagenome Samples 
## Workshop overview
Metagenomics is the application of high-throughput sequencing to DNA extracted directly from environmental, uncultured samples. 
With sequencing becoming more accessible to labs in the Philippines, we are witnessing increasing use of metagenomics to study microbial communities in environmental, ecological, agricultural, epidemiological, and clinical settings.
The power of metagenomics, however, comes with the challenging task of handling and analyzing large volumes of sequence data.
This hands-on workshop will use the case of shotgun metagenomic samples of hospital waste-water to demonstrate bioinformatics workflows for typical tasks such as identifying microbial composition and diversity, making statistical comparisons between samples, detecting presence and diversity of antibiotic resistance genes, etc.

## Instructor
[Anish M.S. Shrestha, Ph.D.](https://www.a-transposable-element.com) \
Head, [Bioinformatics Lab](https://www.bioinfodlsu.com), AdRIC, De La Salle University Manila.

with contributions to the learning material from:
Daphne Go and Paul Yu.

## Target Participants
This workshop would be most beneficial to life scientists and early-career bioinformaticians who have or plan to use metagenomics-based approaches in their research. Prior exposure to Unix command line and R would be helpful. 

## Learning Outcomes
At the end of the workshop, participants are expected to be able to:

- List and distinguish among widely used bioinformatics tools and databases for metagenomic data analysis.
- Identify and apply assembly-free tools for some typical metagaenomic data analysis tasks
- Recognize the role of compositional data analysis

## Learning Plan
The workshop is composed of the following modules:

- Metagenomics data generation process 
- Overview of analysis workflows
- Taxonomic profiling by comparing reads to databases
- Statistical comparison of metagenomic samples
- Identifying presence of antibiotic-resistant genes 

## Learning material
The content is organized as a Jupyter Book, which is a collection of Jupyter Notebooks.
Go to [Getting Started](getting-started.md) for instructions on how to access a copy of the material, data, and software environment.


## Some resources for preparation
- Survey/review articles on metagenomics :
    - Quince et al.: Shotgun metagenomics, from sampling to analysis, Nature Biotechnology, 35:9 (2017).
- A basic tutorial on R can be found [here](https://github.com/bioinfodlsu/basic-r-tutorial).
- [Jupyter Lab Guide](https://jupyterlab.readthedocs.io/en/latest/).




```{tableofcontents}
```
