# Taxonomic Profiling
We begin with the question of figuring out the taxonomic composition of the sampled microbial community, i.e. to find out what species are present and in what proportion. 

There are many tools available for this task. The basic idea behind all of them to assign reads to  sequences in a reference database, and then count the number of reads to compute relative abundance. Broadly speaking, there are two choices for the reference database. 

## Using a database of marker genes known for discriminatory power between species


## Using comprehensive whole-genome databases





## Discussion
What are the (dis)-advantages of the two approaches above? Are they computing the same thing?

## References and further reading
- Ye et al., Benchmarking Metagenomics Tools for Taxonomic Classification, Cell Volume 178, Issue 4, 779 -- 74 (2019)
- Sun et al., Challenges in benchmarking metagenomic profilers, Nature Methods, 18, 618-626 (2021)
