{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ARG Profiling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Indexing the ResFinder Database with Bowtie2\n",
    "The command below uses Bowtie2 to create a searchable index from the ResFinder database's nucleotide sequences. This index is essential for aligning sequence reads to the database during downstream analyses,"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Settings:\n",
      "  Output files: \"/resfinder_db/resfinder.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  /resfinder_db/all.fsa\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "bmax according to bmaxDivN setting: 749187\n",
      "Using parameters --bmax 561891 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 561891 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 374593 (target: 561890)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (561891) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 158836 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 158837 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (561891) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 545749 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 545750 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (561891) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 438356 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 438357 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (561891) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 188863 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 188864 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (561891) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 438971 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 438972 for bucket 5\n",
      "Getting block 6 of 8\n",
      "  Reserving size (561891) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 464910 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 464911 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (561891) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 282543 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 282544 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (561891) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 478513 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 478514 for bucket 8\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 826281\n",
      "fchr[G]: 1505231\n",
      "fchr[T]: 2260845\n",
      "fchr[$]: 2996748\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 5310314 bytes to primary EBWT file: /resfinder_db/resfinder.1.bt2.tmp\n",
      "Wrote 749192 bytes to secondary EBWT file: /resfinder_db/resfinder.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2996748\n",
      "    bwtLen: 2996749\n",
      "    sz: 749187\n",
      "    bwtSz: 749188\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 187297\n",
      "    offsSz: 749188\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 15609\n",
      "    numLines: 15609\n",
      "    ebwtTotLen: 998976\n",
      "    ebwtTotSz: 998976\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 00:00:05\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:00\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:00\n",
      "  Time to reverse reference sequence: 00:00:00\n",
      "bmax according to bmaxDivN setting: 749187\n",
      "Using parameters --bmax 561891 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 561891 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:00:00\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:00\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:00\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 6; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Split 1, merged 1; iterating...\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 428106 (target: 561890)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (561891) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 555057 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 555058 for bucket 1\n",
      "Getting block 2 of 7\n",
      "  Reserving size (561891) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 439495 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 439496 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (561891) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 537700 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 537701 for bucket 3\n",
      "Getting block 4 of 7\n",
      "  Reserving size (561891) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 490567 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:01\n",
      "Returning block of 490568 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (561891) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 347491 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 347492 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (561891) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 429129 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 429130 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (561891) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 197303 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:00:00\n",
      "Returning block of 197304 for bucket 7\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 826281\n",
      "fchr[G]: 1505231\n",
      "fchr[T]: 2260845\n",
      "fchr[$]: 2996748\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 5310314 bytes to primary EBWT file: /resfinder_db/resfinder.rev.1.bt2.tmp\n",
      "Wrote 749192 bytes to secondary EBWT file: /resfinder_db/resfinder.rev.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2996748\n",
      "    bwtLen: 2996749\n",
      "    sz: 749187\n",
      "    bwtSz: 749188\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 187297\n",
      "    offsSz: 749188\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 15609\n",
      "    numLines: 15609\n",
      "    ebwtTotLen: 998976\n",
      "    ebwtTotSz: 998976\n",
      "    color: 0\n",
      "    reverse: 1\n",
      "Total time for backward call to driver() for mirror index: 00:00:04\n",
      "Renaming /resfinder_db/resfinder.3.bt2.tmp to /resfinder_db/resfinder.3.bt2\n",
      "Renaming /resfinder_db/resfinder.4.bt2.tmp to /resfinder_db/resfinder.4.bt2\n",
      "Renaming /resfinder_db/resfinder.1.bt2.tmp to /resfinder_db/resfinder.1.bt2\n",
      "Renaming /resfinder_db/resfinder.2.bt2.tmp to /resfinder_db/resfinder.2.bt2\n",
      "Renaming /resfinder_db/resfinder.rev.1.bt2.tmp to /resfinder_db/resfinder.rev.1.bt2\n",
      "Renaming /resfinder_db/resfinder.rev.2.bt2.tmp to /resfinder_db/resfinder.rev.2.bt2\n"
     ]
    }
   ],
   "source": [
    "!bowtie2-build /resfinder_db/all.fsa /resfinder_db/resfinder"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Mapping Sequencing Reads to the ResFinder Database with Bowtie2\n",
    "This command processes multiple sequencing samples to identify antibiotic resistance genes by aligning them against the ResFinder database using Bowtie2, followed by conversion of the output to BAM format using Samtools. \n",
    "\n",
    "The alignment was performed  using the command bowtie2 -x CARD-db / Resfinder-db -1 reads1 -2 reads2 -D 20 -R 3 -N 1 -L 20 -i S,1,0.50, which corresponds to Bowtie2's \"very sensitive\" preset. This preset was chosen to enhance alignment accuracy by maximizing read capture and achieving a search depth capable of identifying , targeted to increase search depth by capturing up to 95-99% of potential read matches to the reference sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing sample: BFH10_S128\n",
      "3129692 reads; of these:\n",
      "  3129692 (100.00%) were paired; of these:\n",
      "    3129048 (99.98%) aligned concordantly 0 times\n",
      "    69 (0.00%) aligned concordantly exactly 1 time\n",
      "    575 (0.02%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3129048 pairs aligned concordantly 0 times; of these:\n",
      "      21 (0.00%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3129027 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      6258054 mates make up the pairs; of these:\n",
      "        6257400 (99.99%) aligned 0 times\n",
      "        84 (0.00%) aligned exactly 1 time\n",
      "        570 (0.01%) aligned >1 times\n",
      "0.03% overall alignment rate\n",
      "Completed sample: BFH10_S128\n",
      "Processing sample: BFH33_S151\n",
      "3129692 reads; of these:\n",
      "  3129692 (100.00%) were paired; of these:\n",
      "    3127275 (99.92%) aligned concordantly 0 times\n",
      "    540 (0.02%) aligned concordantly exactly 1 time\n",
      "    1877 (0.06%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3127275 pairs aligned concordantly 0 times; of these:\n",
      "      115 (0.00%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3127160 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      6254320 mates make up the pairs; of these:\n",
      "        6252167 (99.97%) aligned 0 times\n",
      "        283 (0.00%) aligned exactly 1 time\n",
      "        1870 (0.03%) aligned >1 times\n",
      "0.12% overall alignment rate\n",
      "Completed sample: BFH33_S151\n",
      "Processing sample: BH02_S77\n",
      "3129692 reads; of these:\n",
      "  3129692 (100.00%) were paired; of these:\n",
      "    3126540 (99.90%) aligned concordantly 0 times\n",
      "    307 (0.01%) aligned concordantly exactly 1 time\n",
      "    2845 (0.09%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3126540 pairs aligned concordantly 0 times; of these:\n",
      "      36 (0.00%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3126504 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      6253008 mates make up the pairs; of these:\n",
      "        6249352 (99.94%) aligned 0 times\n",
      "        445 (0.01%) aligned exactly 1 time\n",
      "        3211 (0.05%) aligned >1 times\n",
      "0.16% overall alignment rate\n",
      "Completed sample: BH02_S77\n",
      "Processing sample: BH03_S78\n",
      "3129692 reads; of these:\n",
      "  3129692 (100.00%) were paired; of these:\n",
      "    3128697 (99.97%) aligned concordantly 0 times\n",
      "    122 (0.00%) aligned concordantly exactly 1 time\n",
      "    873 (0.03%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3128697 pairs aligned concordantly 0 times; of these:\n",
      "      24 (0.00%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3128673 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      6257346 mates make up the pairs; of these:\n",
      "        6256212 (99.98%) aligned 0 times\n",
      "        107 (0.00%) aligned exactly 1 time\n",
      "        1027 (0.02%) aligned >1 times\n",
      "0.05% overall alignment rate\n",
      "Completed sample: BH03_S78\n",
      "Processing sample: FH1_S162\n",
      "3129692 reads; of these:\n",
      "  3129692 (100.00%) were paired; of these:\n",
      "    3125761 (99.87%) aligned concordantly 0 times\n",
      "    845 (0.03%) aligned concordantly exactly 1 time\n",
      "    3086 (0.10%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3125761 pairs aligned concordantly 0 times; of these:\n",
      "      46 (0.00%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3125715 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      6251430 mates make up the pairs; of these:\n",
      "        6249184 (99.96%) aligned 0 times\n",
      "        422 (0.01%) aligned exactly 1 time\n",
      "        1824 (0.03%) aligned >1 times\n",
      "0.16% overall alignment rate\n",
      "Completed sample: FH1_S162\n",
      "Processing sample: FH2_S163\n",
      "3129692 reads; of these:\n",
      "  3129692 (100.00%) were paired; of these:\n",
      "    3128447 (99.96%) aligned concordantly 0 times\n",
      "    537 (0.02%) aligned concordantly exactly 1 time\n",
      "    708 (0.02%) aligned concordantly >1 times\n",
      "    ----\n",
      "    3128447 pairs aligned concordantly 0 times; of these:\n",
      "      77 (0.00%) aligned discordantly 1 time\n",
      "    ----\n",
      "    3128370 pairs aligned 0 times concordantly or discordantly; of these:\n",
      "      6256740 mates make up the pairs; of these:\n",
      "        6255766 (99.98%) aligned 0 times\n",
      "        269 (0.00%) aligned exactly 1 time\n",
      "        705 (0.01%) aligned >1 times\n",
      "0.06% overall alignment rate\n",
      "Completed sample: FH2_S163\n"
     ]
    }
   ],
   "source": [
    "!mkdir -p /home/jovyan/mapped_resfinder\n",
    "\n",
    "!for sample in $(ls /home/jovyan/data/*_R1_*.fastq.gz | sed \"s/.*\\\\///\" | sed \"s/_R1_.*//\"); do \\\n",
    "    echo \"Processing sample: ${sample}\"; \\\n",
    "    bowtie2 -x /resfinder_db/resfinder \\\n",
    "        -1 /home/jovyan/data/${sample}_R1_trimmed.fastq.gz \\\n",
    "        -2 /home/jovyan/data/${sample}_R2_trimmed.fastq.gz \\\n",
    "        -D 20 -R 3 -N 1 -L 20 -i S,1,0.50 | \\\n",
    "    samtools view -Sb - > /home/jovyan/mapped_resfinder/${sample}_unfiltered.bam 2>&1; \\\n",
    "    echo \"Completed sample: $sample\"; \\\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Sorting Reads by Position: \n",
    "\n",
    "The command sorts the reads in the BAM file by their position on the reference genome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing sample: BFH10_S128\n",
      "[bam_sort_core] merging from 2 files and 1 in-memory blocks...\n",
      "Processing sample: BFH33_S151\n",
      "[bam_sort_core] merging from 2 files and 1 in-memory blocks...\n",
      "Processing sample: BH02_S77\n",
      "[bam_sort_core] merging from 2 files and 1 in-memory blocks...\n",
      "Processing sample: BH03_S78\n",
      "[bam_sort_core] merging from 2 files and 1 in-memory blocks...\n",
      "Processing sample: FH1_S162\n",
      "[bam_sort_core] merging from 2 files and 1 in-memory blocks...\n",
      "Processing sample: FH2_S163\n",
      "[bam_sort_core] merging from 2 files and 1 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "!mkdir -p /home/jovyan/sorted_reads_resfinder\n",
    "\n",
    "!for sample in $(ls /home/jovyan/data/*_R1_*.fastq.gz | sed \"s/.*\\\\///\" | sed \"s/_R1_.*//\"); do \\\n",
    "    echo \"Processing sample: ${sample}\"; \\\n",
    "    samtools sort -T /home/jovyan/sorted_reads_resfinder/${sample} -O bam \\\n",
    "    -o /home/jovyan/sorted_reads_resfinder/${sample}.bam \\\n",
    "    /home/jovyan/mapped_resfinder/${sample}_unfiltered.bam; \\\n",
    "done\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Indexing Sorted BAM Files\n",
    "This command processes the sorted BAM files generated previously and creates index files for them using Samtools. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing sample: BFH10_S128\n",
      "Processing sample: BFH33_S151\n",
      "Processing sample: BH02_S77\n",
      "Processing sample: BH03_S78\n",
      "Processing sample: FH1_S162\n",
      "Processing sample: FH2_S163\n"
     ]
    }
   ],
   "source": [
    "!for sample in $(ls /home/jovyan/data/*_R1_*.fastq.gz | sed 's/.*\\///' | sed 's/_R1_.*//'); do \\\n",
    "    echo \"Processing sample: ${sample}\"; \\\n",
    "    samtools index /home/jovyan/sorted_reads_resfinder/${sample}.bam;\\\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extracting Gene Names and Count of Mapped Reads Using samtools idxstats\n",
    "This command extracts gene names (or reference names) from the sorted BAM files using samtools idxstats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing sample: BFH10_S128\n",
      "Processing sample: BFH33_S151\n",
      "Processing sample: BH02_S77\n",
      "Processing sample: BH03_S78\n",
      "Processing sample: FH1_S162\n",
      "Processing sample: FH2_S163\n"
     ]
    }
   ],
   "source": [
    "!mkdir -p /home/jovyan/results\n",
    "!for sample in $(ls /home/jovyan/data/*_R1_*.fastq.gz | sed 's/.*\\///' | sed 's/_R1_.*//'); do \\\n",
    "    echo \"Processing sample: ${sample}\";\\\n",
    "    samtools idxstats /home/jovyan/sorted_reads_resfinder/${sample}.bam | grep -v \"\\*\" | cut -f1 | \\\n",
    "    (echo \"GENE\"; cat) > /home/jovyan/results/gene_names;\\\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using samtools to get the counts of mapped reads for each reference gene from the aligned BAM file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing sample: BFH10_S128\n",
      "Processing sample: BFH33_S151\n",
      "Processing sample: BH02_S77\n",
      "Processing sample: BH03_S78\n",
      "Processing sample: FH1_S162\n",
      "Processing sample: FH2_S163\n"
     ]
    }
   ],
   "source": [
    "!for sample in $(ls /home/jovyan/data/*_R1_*.fastq.gz | sed 's/.*\\///' | sed 's/_R1_.*//'); do \\\n",
    "    echo \"Processing sample: ${sample}\"; \\\n",
    "    echo -n \"${sample}\" > /home/jovyan/results/${sample}_counts; \\\n",
    "    samtools idxstats /home/jovyan/sorted_reads_resfinder/${sample}.bam | grep -v \"\\\\*\" | cut -f3 >> /home/jovyan/results/${sample}_counts; \\\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Combine into one file\n",
    "This command combines the count of mapped reads from all samples into a single tab-delimited file, where each row represents a gene and each column represents a sample. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "!echo -e \"GENE\\t$(ls /home/jovyan/data/*_R1_*.fastq.gz | sed 's/.*\\///' | sed 's/_R1_.*//' | tr '\\n' '\\t' | sed 's/\\t$//')\" > /home/jovyan/results/ARG_genemat.txt\n",
    "\n",
    "!paste /home/jovyan/results/gene_names $(ls /home/jovyan/data/*_R1_*.fastq.gz | \\\n",
    "sed 's/.*\\///' | sed 's/_R1_.*//' | sed 's|^|/home/jovyan/results/|g' | sed 's|$|_counts|') |\\\n",
    "tail -n +2 >> /home/jovyan/results/ARG_genemat.txt\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GENE\tBFH10_S128\tBFH33_S151\tBH02_S77\tBH03_S78\tFH1_S162\tFH2_S163\n",
      "aac(6')-Ib_2_M23634\t1\t3\t14\t0\t49\t5\n",
      "aac(6')-Ib11_1_AY136758\t6\t22\t21\t4\t101\t50\n",
      "aac(6')-30-aac(6')-Ib'_1_AJ584652\t0\t0\t0\t0\t0\t0\n",
      "aac(6')-Iaj_1_AB709942\t0\t0\t0\t0\t0\t0\n",
      "aac(6')-Ian_1_AP014611\t0\t0\t0\t0\t0\t0\n",
      "aac(6')-Iak_1_AB894482\t4\t16\t10\t1\t86\t21\n",
      "aac(6')-Ib-Hangzhou_1_FJ503047\t0\t0\t0\t0\t0\t0\n",
      "aac(6')-Iid_1_AJ584700\t0\t0\t0\t0\t0\t0\n",
      "aac(6')-Iih_1_AJ584701\t0\t0\t0\t0\t0\t0\n"
     ]
    }
   ],
   "source": [
    "!head /home/jovyan/results/ARG_genemat.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Metaxa2: Improved Identification and Taxonomic Classification of Small and Large Subunit rRNA in Metagenomic Data\n",
    "- Used for normalization of the counts of mapped reads later on."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "For each sample, the script runs Metaxa2, a tool for detecting ribosomal RNA (rRNA) in metagenomic data with the following parameters:\n",
    "\n",
    "- -1 and -2 specify the paired-end read files (forward and reverse)\n",
    "- --align none disables alignment (typically used when only taxonomic classification is desired).\n",
    "- --graphical F disables graphical output (if you don't want plots).\n",
    "- --plus tells Metaxa2 to use the BLAST+ for alignment.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metaxa Processing Sample BFH10_S128\n",
      "Metaxa Processing Sample BFH33_S151\n",
      "Metaxa Processing Sample BH02_S77\n",
      "Metaxa Processing Sample BH03_S78\n",
      "Metaxa Processing Sample FH1_S162\n",
      "Metaxa Processing Sample FH2_S163\n"
     ]
    }
   ],
   "source": [
    "!mkdir -p /home/jovyan/metaxa2\n",
    "!for sample in $(ls /home/jovyan/data/*_R1_*.fastq.gz | sed 's/.*\\///' | sed 's/_R1_.*//'); do \\\n",
    "    echo \"Metaxa Processing Sample $sample\";\\\n",
    "    metaxa2 -1 /home/jovyan/data/${sample}_R1_trimmed.fastq.gz -2 /home/jovyan/data/${sample}_R2_trimmed.fastq.gz \\\n",
    "    -f fastq -z gzip -t b -o /home/jovyan/metaxa2/${sample} --align none --graphical F --plus 2>/dev/null;\\\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Processing Metaxa Output:\n",
    "\n",
    "The command metaxa2_ttt processes the taxonomy output from each sample, extracting counts of identified taxa (e.g., genus-level counts) for each sample's rRNA sequences.\n",
    "This generates raw counts of each identified genus in each sample.\n",
    "Combining Counts at the Genus Level with metaxa2_dc:\n",
    "\n",
    "The metaxa2_dc step combines these raw taxonomy counts across all samples at the genus level.\n",
    "It generates a summary file (metaxa_genus.txt) that includes the counts of each genus detected in every sample.\n",
    "Normalization Using Genus-Level Counts:\n",
    "\n",
    "The genus-level counts produced by metaxa2_dc can be used as input for normalization. Since sequencing depth varies across samples, normalization ensures that the observed counts are comparable.\n",
    "Normalization involves scaling counts to a common factor (e.g., total reads, sequencing depth, or per million reads)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metaxa Taxonomic Traversal Tool -- Processes Taxonomic Output From Metaxa\n",
      "by Johan Bengtsson-Palme, University of Gothenburg\n",
      "Version: 2.2\n",
      "-----------------------------------------------------------------\n",
      "Metaxa Taxonomic Traversal Tool -- Processes Taxonomic Output From Metaxa\n",
      "by Johan Bengtsson-Palme, University of Gothenburg\n",
      "Version: 2.2\n",
      "-----------------------------------------------------------------\n",
      "Metaxa Taxonomic Traversal Tool -- Processes Taxonomic Output From Metaxa\n",
      "by Johan Bengtsson-Palme, University of Gothenburg\n",
      "Version: 2.2\n",
      "-----------------------------------------------------------------\n",
      "Metaxa Taxonomic Traversal Tool -- Processes Taxonomic Output From Metaxa\n",
      "by Johan Bengtsson-Palme, University of Gothenburg\n",
      "Version: 2.2\n",
      "-----------------------------------------------------------------\n",
      "Metaxa Taxonomic Traversal Tool -- Processes Taxonomic Output From Metaxa\n",
      "by Johan Bengtsson-Palme, University of Gothenburg\n",
      "Version: 2.2\n",
      "-----------------------------------------------------------------\n",
      "Metaxa Taxonomic Traversal Tool -- Processes Taxonomic Output From Metaxa\n",
      "by Johan Bengtsson-Palme, University of Gothenburg\n",
      "Version: 2.2\n",
      "-----------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!for sample in $(ls /home/jovyan/data/*_R1_*.fastq.gz | sed 's/.*\\///' | sed 's/_R1_.*//'); do \\\n",
    "    metaxa2_ttt -i /home/jovyan/metaxa2/${sample}.taxonomy.txt -t b -o /home/jovyan/metaxa2/${sample};\\\n",
    "done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Metaxa2 Diversity Tools - Data Collector\n",
      "by Johan Bengtsson-Palme, University of Gothenburg\n",
      "Version: 2.2\n",
      "This program is distributed under the GNU GPL 3 license, use the --license option for more information on this license.\n",
      "-----------------------------------------------------------------\n"
     ]
    }
   ],
   "source": [
    "!mkdir -p /home/jovyan/results\n",
    "!metaxa2_dc -o /home/jovyan/results/metaxa_genus.txt /home/jovyan/metaxa2/*.level_6.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reformat Sample Name Headers to Exclude File Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "!sed -i '1s|/home/jovyan/metaxa2/||g' /home/jovyan/results/metaxa_genus.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Taxa\tBFH10_S128\tBFH33_S151\tBH02_S77\tBH03_S78\tFH1_S162\tFH2_S163\n",
      "Bacteria;Acidobacteria;Acidobacteria;Acidobacteriales;Acidobacteriaceae;Unclassified Acidobacteriaceae\t0\t0\t4\t0\t0\t0\n",
      "Bacteria;Acidobacteria;Acidobacteria;Acidobacteriales;Unclassified Acidobacteriales;\t0\t1\t0\t0\t0\t0\n",
      "Bacteria;Acidobacteria;Acidobacteria;DA023;Unclassified DA023;\t1\t0\t0\t1\t0\t0\n",
      "Bacteria;Acidobacteria;Acidobacteria;Order Incertae Sedis;Family Incertae Sedis;Bryobacter\t0\t0\t0\t1\t0\t0\n",
      "Bacteria;Acidobacteria;Acidobacteria;Unclassified Acidobacteria;;\t3\t0\t5\t0\t0\t0\n",
      "Bacteria;Acidobacteria;Holophagae;Holophagales;Holophagaceae;Geothrix\t0\t1\t7\t0\t0\t0\n",
      "Bacteria;Acidobacteria;Holophagae;Holophagales;Holophagaceae;Holophaga\t0\t0\t1\t0\t1\t0\n",
      "Bacteria;Acidobacteria;Holophagae;Holophagales;Holophagaceae;Unclassified Holophagaceae\t0\t0\t4\t0\t0\t0\n",
      "Bacteria;Acidobacteria;Unclassified Acidobacteria;;;\t0\t1\t27\t0\t0\t0\n"
     ]
    }
   ],
   "source": [
    "!head /home/jovyan/results/metaxa_genus.txt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
